### ISSUE TITLE
[SQLITE_BUSY]  The database file is locked (database is locked)

### ISSUE THREAD

**Issue by jayaramcs** [npgall/cqengine#227]
Hi Niall

I am running into this in a multithreaded environment when I add an item to the index. 
 

```
! org.sqlite.SQLiteException: [SQLITE_BUSY]  The database file is locked (database is locked)
! at org.sqlite.core.DB.newSQLException(DB.java:909)
! at org.sqlite.core.DB.newSQLException(DB.java:921)
! at org.sqlite.core.DB.throwex(DB.java:886)
! at org.sqlite.core.DB.executeBatch(DB.java:774)
! at org.sqlite.core.CorePreparedStatement.executeBatch(CorePreparedStatement.java:84)
! at com.googlecode.cqengine.index.sqlite.support.DBQueries.bulkAdd(DBQueries.java:302)
! ... 15 common frames omitted
! Causing: java.lang.IllegalStateException: Unable to bulk add rows to the index table: Id. Rolled back: true
! at com.googlecode.cqengine.index.sqlite.support.DBQueries.bulkAdd(DBQueries.java:320)
! at com.googlecode.cqengine.index.sqlite.SQLiteIndex.doAddAll(SQLiteIndex.java:468)
! at com.googlecode.cqengine.index.sqlite.SQLiteIndex.addAll(SQLiteIndex.java:425)
! at com.googlecode.cqengine.index.sqlite.SQLiteIdentityIndex.addAll(SQLiteIdentityIndex.java:113)
! at com.googlecode.cqengine.persistence.support.sqlite.SQLiteObjectStore.add(SQLiteObjectStore.java:106)
! at com.googlecode.cqengine.ConcurrentIndexedCollection.add(ConcurrentIndexedCollection.java:350)
```

Index is setup as follows (in a singleton class) 

```
private final IndexedCollection<PendingItem> pendingItems  = new ConcurrentIndexedCollection<PendingItem>(DiskPersistence.onPrimaryKeyInFile(PendingItem.ID, persistentIndex)); 
```
On calling  ``` pendingItems.add(item);```  I run into the DB Locked error. 

Any ideas ? Is add not threadsafe?

**Comment by npgall** [comment]
This might happen if reading threads are not closing the `ResultSet`s after use. Could you check that?

**Comment by jayaramcs** [comment]
I have used 'try with resources' for all the query and couldnt spot any obvious errors. 

My current workaround (not included here) is to use ReentrantReadWriteLock so that reads and writes are not stepping on each other. Still testing this out to see if it resolves the DB locking. 

Example 
```java
    public boolean isAwaitingResultsLookup(String lookupId) {
         try (ResultSet<PendingItem> results
                = pendingItems.retrieve(equal(PendingItem.LOOKUP_ID, lookupId))) {
            return results.stream().findAny().isPresent();
        }
    }
     
    public void addItem(PendingItem item) {
        if (!isAwaitingResultsLookup(item.getSessionId(), item.getLookupId())) {
            pendingItems.add(item);
        }
    } 
```

The implementation itself is fairly small, here is the complete (scrubed) code. 


```java
public class PendingItemsMgr {

    private final IndexedCollection<PendingItem> pendingItems;
 
    public List<PendingItem> getAllItems() {
        
        List<PendingItem> allPendingItems = new ArrayList<>();
        
        try(ResultSet<PendingItem> results = pendingItems.retrieve(all(PendingItem.class))){
            results.forEach(item -> allPendingItems.add(item));
        }
        
        return allPendingItems;
    }


    public List<PendingItem> getWaitingItems() {
        
        List<PendingItem> allPendingItems = new ArrayList<>();
        
        try(ResultSet<PendingItem> results = pendingItems.retrieve(lessThanOrEqualTo(PendingItem.LOOKUP_TS, Instant.now().getEpochSecond()))){
            results.forEach(item -> allPendingItems.add(item));
        }
        
        return allPendingItems;
    }
    
    public long getWaitingItemsCount(String sessionId) {
        try (ResultSet<PendingItem> results
                = pendingItems.retrieve(
                         equal(PendingItem.SESSION_ID, sessionId))) {
            return results.stream().count();
        }
    }
    
    public boolean isAwaitingResultsLookup(String sessionId, String lookupId) {
         try (ResultSet<PendingItem> results
                = pendingItems.retrieve(
                        and(equal(PendingItem.SESSION_ID, sessionId), equal(PendingItem.LOOKUP_ID, lookupId)))) {
            return results.stream().findAny().isPresent();
        }
    }
    
    public boolean isAwaitingResultsLookup(String lookupId) {
         try (ResultSet<PendingItem> results
                = pendingItems.retrieve(equal(PendingItem.LOOKUP_ID, lookupId))) {
            return results.stream().findAny().isPresent();
        }
    }
     
    public void addItem(PendingItem item) {
        if (!isAwaitingResultsLookup(item.getSessionId(), item.getLookupId())) {
            pendingItems.add(item);
        }
    } 
  
    public void removeItem(String sessionId, String lookupId) {
        try (ResultSet<PendingItem> results
                = pendingItems.retrieve(
                        and(equal(PendingItem.SESSION_ID, sessionId), equal(PendingItem.LOOKUP_ID, lookupId)))) {

            results.forEach(item -> pendingItems.remove(item));
        }
    }
    

    private PendingItemsMgr() {


        File persistentIndex = new File("pending.dat");

        pendingItems
                = new ConcurrentIndexedCollection<PendingItem>(DiskPersistence.onPrimaryKeyInFile(PendingItem.ID, persistentIndex));
        
        pendingItems.addIndex(CompoundIndex.onAttributes(PendingItem.SESSION_ID, PendingItem.LOOKUP_ID));
        pendingItems.addIndex(NavigableIndex.onAttribute(PendingItem.LOOKUP_TS));
        
    }

    public static PendingItemsMgr get() {
        return HOLDER.INSTANCE;
    }

    private static class HOLDER {

        static PendingItemsMgr INSTANCE = new PendingItemsMgr();
    }
}```

**Comment by codingchili** [comment]
Ahoy, I have the same issue.

Reproducer; fails with the same exception in 90% of runs. 
```java
    private static final AtomicInteger idCounter = new AtomicInteger(0);

    public class MyTestClass {
        Integer id = idCounter.incrementAndGet();
    }

    @Test
    public void writeConcurrencyAndDiskPersistence() throws InterruptedException {
        SimpleAttribute<MyTestClass, Integer> attribute = new SimpleAttribute<MyTestClass, Integer>() {
            @Override
            public Integer getValue(MyTestClass object, QueryOptions queryOptions) {
                return object.id;
            }
        };

        IndexedCollection<MyTestClass> collection = new ConcurrentIndexedCollection<>(
                DiskPersistence.onPrimaryKeyInFile(attribute, DiskPersistence.createTempFile())
        );

        for (int i = 0; i < 2; i++) {
            new Thread(() -> collection.add(new MyTestClass())).start();
        }
        Thread.sleep(2000);
    }
```
Copied and pasted into DiskPersistenceConcurrencyTest on the master branch.

Output;
```
Exception in thread "Thread-1" java.lang.IllegalStateException: Unable to bulk add rows to the index table: UnnamedattributeclasscomgooglecodecqenginepersistencediskDiskPersistenceConcurrencyTest1. Rolled back: true
	at com.googlecode.cqengine.index.sqlite.support.DBQueries.bulkAdd(DBQueries.java:320)
	at com.googlecode.cqengine.index.sqlite.SQLiteIndex.doAddAll(SQLiteIndex.java:468)
	at com.googlecode.cqengine.index.sqlite.SQLiteIndex.addAll(SQLiteIndex.java:425)
	at com.googlecode.cqengine.index.sqlite.SQLiteIdentityIndex.addAll(SQLiteIdentityIndex.java:113)
	at com.googlecode.cqengine.persistence.support.sqlite.SQLiteObjectStore.add(SQLiteObjectStore.java:106)
	at com.googlecode.cqengine.ConcurrentIndexedCollection.add(ConcurrentIndexedCollection.java:350)
	at com.googlecode.cqengine.persistence.disk.DiskPersistenceConcurrencyTest.lambda$writeConcurrencyAndDiskPersistence$0(DiskPersistenceConcurrencyTest.java:65)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: org.sqlite.SQLiteException: [SQLITE_BUSY]  The database file is locked (database is locked)
	at org.sqlite.core.DB.newSQLException(DB.java:909)
	at org.sqlite.core.DB.newSQLException(DB.java:921)
	at org.sqlite.core.DB.throwex(DB.java:886)
	at org.sqlite.core.DB.executeBatch(DB.java:774)
	at org.sqlite.core.CorePreparedStatement.executeBatch(CorePreparedStatement.java:84)
	at com.googlecode.cqengine.index.sqlite.support.DBQueries.bulkAdd(DBQueries.java:302)
	... 7 more
```
Javadocs on both ConcurrentIndexedCollection and DiskPersistence states that they support "concurrent reads and writes". But it seems like DiskPersistence does not support concurrent writes?

With serialized writes it works and I get around **70** insert ops per second, with only the implicit index on the primary key.

**Comment by codingchili** [comment]
Some more observations
- ConcurrentIndexedCollection and ObjectLockingIndexedCollection has the same issue.
- TransactionalIndexedCollection does **not** have the same issue.
- transaction mode EXCLUSIVE or IMMEDIATE does **not** result in the issue.
- affects both: SQLiteIndex doAddAll/removeAll.

Sync mode NORMAL does not result in the issue, but only if `DBQueries.setSyncAndJournaling(connection, pragmaSynchronous, pragmaJournalMode);` is called just before the `DBQueries.bulkAdd/bulkRemove` - if it's called before `createIndexOnTable` it does not seem to apply. It needs to be called even if the getPragma methods returns the expected values (WAL, NORMAL).

For WAL it appears safe to use NORMAL sync.
https://www.sqlite.org/pragma.html#pragma_synchronous

The problem is that it's not enough to set the sync property on the properties object passed to CQEngine, the additional calls to `setSyncAndJournaling` are required. I'm not having any concurrency issues with the changes in the #229. All my tests are passing, have not run all CQEngine test cases yet. 

While insert performance is increased, query performance is completely killed. That might be why we are defaulting to FULL sync. 

I have not used sqlite before, hopefully someone else could provide some more insight into the matter.

**Comment by npgall** [comment]
Nice work deep diving there @codingchili and @jayaramcs !

I am hoping to get some time in the next week (actually next week) to investigate this.
Your test cases and findings above will make it a lot easier to get to the bottom of this.

I am also hoping to experiment with changing how DiskPersistence (does not) cache a connection to prevent the database from getting closed/reopened, in the same way that OnHeapPersistence does. This might actually improve performance and could solve this issue where the file is getting locked at SQLite level.

**Comment by jayaramcs** [comment]
Happy to rerun my tests if that helps.

**Comment by npgall** [comment]
I've merged @codingchili 's PR - many thanks! Those changes are now enabled by default - and they seem to work well for me.

I also added another feature to enable SQLite's Shared Cache Mode (disabled by default).
In my testing, for some high throughput but low concurrency workloads, the shared cache doubled the TPS.
However in other workloads with a lot of concurrency, the shared cache caused database busy errors. It does not support concurrent writing threads well. Could you both let me know if the latest code works for you?

**Comment by codingchili** [comment]
Nice! I will run my test cases again sometime later this week and report back.

**Comment by npgall** [comment]
I have now released these changes as CQEngine 3.1.0, so you can just try that version. Thanks!

**Comment by npgall** [comment]
I will close this now as I think the issue has been resolved. Thanks again, and feel free to reopen if you still have issues with the latest version!
